modules/ldap_query/lambda/lambda.py

```diff
@@ -0,0 +1,443 @@
+        """Creates a ldif document with the query results"""
+        # could be an alternative way of user disablement
+
+    @staticmethod
+    def byte_decode_search_results(search_results):
+        users = []
+        for user in search_results:
+            # if the dn is None, skip it.
+            if not user[1][0]:
+                continue
+            else:
+                user_obj = {}
+                for attribute in user[1][1]:
+                    if "ldap://" not in attribute:
+                        user[1][1][attribute] = [
+                            item.decode(encoding="utf-8", errors="ignore")
+                            for item in user[1][1][attribute]
+                        ]
+                user_obj["dn"] = user[1][0]
+                user_obj["user"] = user[1][1]
+                users.append(user_obj)
+        return users
+
+    def is_special(self, sam_name, uac):
+        # list of accounts not to touch
+        disabled_codes = [
+            "514",  # Disabled Account
+            "65536",  # DONT_EXPIRE_PASSWORD
+            "66048",  # Enabled, Password Doesn’t Expire
+            "66050",  # Disabled, Password Doesn’t Expire
+            "66080",  # Disabled, Password Doesn’t Expire & Not Required
+            "262658",  # Disabled, Smartcard Required
+            "262690",  # Disabled, Smartcard Required, Password Not Required
+        ]
+        return uac in disabled_codes or fnmatch.filter(self.filter_patterns, sam_name)
+
+    @staticmethod
+    def filetime_to_dt(ft):
+        """
+        Convert windowsfiletime to python datetime.
+        ref: https://gist.github.com/Mostafa-Hamdy-Elgiar/9714475f1b3bc224ea063af81566d873  # noqa: E501
+        """
+        # January 1, 1970 as MS file time
+        epoch_as_filetime = 116444736000000000
+        hundreds_of_nanoseconds = 10000000
+        return datetime.utcfromtimestamp(
+            (int(ft) - epoch_as_filetime) / hundreds_of_nanoseconds
+        )
+
+
+def get_file_name(file_name, extension):
+    timestamp = datetime.now().strftime("%Y_%m_%d_T%H%M%S.%f")
+    return f"{file_name}_{timestamp}.{extension}"
+
+
+def create_json_doc(**content):
+    """create a table"""
+    # This can be fleshed out to make the retrieved information
+    # more user friendly if desired/required
+    artifact = {}
+    artifact["content"] = json.dumps(content["users"])
+    artifact["file_name"] = get_file_name("user_expiration_table", "json")
+    artifact["raw_scan_results"] = True
+    return artifact
+
+
+def get_html_table_headers(**content):
+    """
+    Return the keys of the first element in the user_list dict.
+
+    We don't care if the dict is unordered b/c the expectation is
+    that each element of the dict will be structured the same
+    """
+    table_headers = []
+    user_list = content["users"]
+    days_since_pwdlastset = content["days_since_pwdlastset"]
+    try:
+        for key in user_list[days_since_pwdlastset][0].keys():
```

Comment:
why iterate? also don't need the `table_headers` var...
```suggestion
        return user_list[days_since_pwdlastset][0].keys()
```

Reply:
updated